File: crates/hyle/src/query.rs
use indexmap::IndexMap;
use serde::{Deserialize, Serialize};
use serde_json::Value as JsonValue;

// ── Mutation input ─────────────────────────────────────────────────────────────

/// The input passed to a mutation (create, update, or delete).
///
/// `model` is the model name (mirrors `Query.model`), allowing generic adapters
/// to route to the correct endpoint without knowing the model at compile time.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
pub struct MutateInput {
    /// The model name — e.g. `"user"`.
    pub model: String,
    /// The row id when editing or deleting an existing record; `None` for creates.
    pub id: Option<JsonValue>,
    /// The form field values to submit.
    pub data: IndexMap<String, String>,
}

// ── URL query string parsing ───────────────────────────────────────────────────

/// Parse a URL query string into `(page, per_page, filters)`.
///
/// - `page` defaults to `1`; `per_page` defaults to `default_per_page`.
/// - Keys `"page"` and `"per_page"` are consumed; all others go into `filters`.
/// - Empty-value keys are omitted from `filters`.
/// - Percent-encoding and `+`-as-space are decoded.
///
/// # Example
/// ```
/// let (page, per_page, filters) = hyle::parse_query_params("page=2&name=Ali", 5);
/// assert_eq!(page, 2);
/// assert_eq!(per_page, 5);
/// assert_eq!(filters["name"], "Ali");
/// ```
pub fn parse_query_params(
    query_str: &str,
    default_per_page: usize,
) -> (usize, usize, IndexMap<String, String>) {
    let mut page = 1usize;
    let mut per_page = default_per_page;
    let mut filters: IndexMap<String, String> = IndexMap::new();

    for part in query_str.split('&').filter(|s| !s.is_empty()) {
        let mut kv = part.splitn(2, '=');
        let k = match kv.next() {
            Some(k) => k,
            None => continue,
        };
        let v = percent_decode(kv.next().unwrap_or(""));
        match k {
            "page"     => { if let Ok(n) = v.parse::<usize>() { page     = n.max(1); } }
            "per_page" => { if let Ok(n) = v.parse::<usize>() { per_page = n.max(1); } }
            _ => {
                if !v.is_empty() {
                    // Repeated keys (e.g. tags=rust&tags=web) are joined with commas
                    // so that filter_rows can match against the comma-joined value.
                    filters
                        .entry(k.to_owned())
                        .and_modify(|existing| {
                            existing.push(',');
                            existing.push_str(&v);
                        })
                        .or_insert(v);
                }
            }
        }
    }

    (page, per_page, filters)
}

fn percent_decode(s: &str) -> String {
    let s = s.replace('+', " ");
    let bytes = s.as_bytes();
    let mut out: Vec<u8> = Vec::with_capacity(bytes.len());
    let mut i = 0;
    while i < bytes.len() {
        if bytes[i] == b'%' && i + 2 < bytes.len() {
            if let Ok(byte) = u8::from_str_radix(
                std::str::from_utf8(&bytes[i + 1..i + 3]).unwrap_or(""),
                16,
            ) {
                out.push(byte);
                i += 3;
                continue;
            }
        }
        out.push(bytes[i]);
        i += 1;
    }
    String::from_utf8_lossy(&out).into_owned()
}

#[derive(Clone, Debug, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Sort {
    pub field: String,
    #[serde(default)]
    pub ascending: bool,
}

#[derive(Clone, Debug, Default, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Query {
    pub model: String,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub select: Vec<String>,
    #[serde(rename = "where", default, skip_serializing_if = "IndexMap::is_empty")]
    pub where_: IndexMap<String, JsonValue>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub filters: Vec<Vec<String>>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub page: Option<usize>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub per_page: Option<usize>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub sort: Option<Sort>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub method: Option<String>,
}

impl Query {
    pub fn new(model: impl Into<String>) -> Self {
        Self {
            model: model.into(),
            ..Self::default()
        }
    }

    pub fn select(mut self, fields: impl IntoIterator<Item = impl Into<String>>) -> Self {
        self.select = fields.into_iter().map(Into::into).collect();
        self
    }

    pub fn filter_layout<I, J, S>(mut self, rows: I) -> Self
    where
        I: IntoIterator<Item = J>,
        J: IntoIterator<Item = S>,
        S: Into<String>,
    {
        self.filters = rows
            .into_iter()
            .map(|row| row.into_iter().map(Into::into).collect())
            .collect();
        self
    }

    pub fn where_eq(mut self, field: impl Into<String>, value: JsonValue) -> Self {
        self.where_.insert(field.into(), value);
        self
    }

    pub fn page(mut self, page: usize, per_page: usize) -> Self {
        self.page = Some(page);
        self.per_page = Some(per_page);
        self
    }

    pub fn sort_by(mut self, field: impl Into<String>, ascending: bool) -> Self {
        self.sort = Some(Sort {
            field: field.into(),
            ascending,
        });
        self
    }
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Manifest {
    pub base: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub id: Option<JsonValue>,
    pub fields: Vec<String>,
    #[serde(default, skip_serializing_if = "IndexMap::is_empty")]
    pub filter: IndexMap<String, JsonValue>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub lookups: Vec<String>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub inlines: Vec<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub page: Option<usize>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub per_page: Option<usize>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub sort: Option<Sort>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub method: Option<String>,
    /// 2-D filter layout carried from `query.filters` for UI layout derivation.
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub filter_fields: Vec<Vec<String>>,
}

// ── Tests ─────────────────────────────────────────────────────────────────────

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parse_empty_string() {
        let (page, per_page, filters) = parse_query_params("", 5);
        assert_eq!(page, 1);
        assert_eq!(per_page, 5);
        assert!(filters.is_empty());
    }

    #[test]
    fn parse_page_and_per_page() {
        let (page, per_page, filters) = parse_query_params("page=3&per_page=10", 5);
        assert_eq!(page, 3);
        assert_eq!(per_page, 10);
        assert!(filters.is_empty());
    }

    #[test]
    fn parse_page_zero_clamps_to_one() {
        let (page, _, _) = parse_query_params("page=0", 5);
        assert_eq!(page, 1);
    }

    #[test]
    fn parse_filters() {
        let (page, per_page, filters) = parse_query_params("name=Ali&role=admin", 5);
        assert_eq!(page, 1);
        assert_eq!(per_page, 5);
        assert_eq!(filters["name"], "Ali");
        assert_eq!(filters["role"], "admin");
    }

    #[test]
    fn parse_empty_value_omitted_from_filters() {
        let (_, _, filters) = parse_query_params("name=", 5);
        assert!(!filters.contains_key("name"));
    }

    #[test]
    fn parse_percent_encoding() {
        let (_, _, filters) = parse_query_params("name=Ali%20Smith", 5);
        assert_eq!(filters["name"], "Ali Smith");
    }

    #[test]
    fn parse_plus_as_space() {
        let (_, _, filters) = parse_query_params("name=Ali+Smith", 5);
        assert_eq!(filters["name"], "Ali Smith");
    }

    #[test]
    fn parse_default_per_page_respected() {
        let (_, per_page, _) = parse_query_params("page=2", 20);
        assert_eq!(per_page, 20);
    }

    #[test]
    fn parse_percent_encoding_multibyte_utf8() {
        // "héllo" encoded: é = %C3%A9
        let (_, _, filters) = parse_query_params("name=h%C3%A9llo", 5);
        assert_eq!(filters["name"], "héllo");
    }

    #[test]
    fn parse_repeated_key_joins_with_comma() {
        // Repeated URL params (e.g. ?tags=rust&tags=web from checkboxes) must be
        // joined with a comma so filter_rows can split them back.
        let (_, _, filters) = parse_query_params("tags=rust&tags=web", 5);
        assert_eq!(filters["tags"], "rust,web");
    }

    #[test]
    fn parse_single_array_key_not_joined() {
        // A single occurrence of a key is stored as-is (no trailing comma).
        let (_, _, filters) = parse_query_params("tags=rust", 5);
        assert_eq!(filters["tags"], "rust");
    }

    #[test]
    fn where_serialises_as_where_not_where_underscore() {
        use serde_json::json;
        let q = Query::new("user").where_eq("name", json!("Alice"));
        let serialised = serde_json::to_value(&q).unwrap();
        assert!(serialised.get("where").is_some(), "expected 'where' key");
        assert!(serialised.get("where_").is_none(), "unexpected 'where_' key");
        let round_trip: Query = serde_json::from_value(serialised).unwrap();
        assert_eq!(round_trip.where_.get("name"), Some(&json!("Alice")));
    }
}

File: zig-bindings/src/query.zig
const std = @import("std");
const raw = @import("raw.zig");
const Value = raw.Value;
const StringHashMap = raw.StringHashMap;

pub const Sort = struct {
    field: []const u8,
    ascending: bool,
};

pub const Query = struct {
    model: []const u8,
    select: [][]const u8,
    where_: StringHashMap(Value),
    filters: [][][]const u8,
    page: ?usize,
    per_page: ?usize,
    sort: ?Sort,
    method: ?[]const u8,

    pub fn init(model: []const u8, allocator: std.mem.Allocator) Query {
        return .{
            .model = model,
            .select = &.{},
            .where_ = StringHashMap(Value).init(allocator),
            .filters = &.{},
            .page = null,
            .per_page = null,
            .sort = null,
            .method = null,
        };
    }

    pub fn withSelect(self: *Query, fields: []const []const u8, allocator: std.mem.Allocator) void {
        self.select = tryAllocSlice(allocator, fields);
    }

    pub fn withWhere(self: *Query, field: []const u8, value: Value) void {
        self.where_.put(field, value) catch {};
    }

    pub fn withFilterLayout(self: *Query, rows: []const []const []const u8, allocator: std.mem.Allocator) void {
        _ = allocator;
        self.filters = rows;
    }

    pub fn withPage(self: *Query, page: usize, per_page: usize) void {
        self.page = page;
        self.per_page = per_page;
    }

    pub fn withSort(self: *Query, field: []const u8, ascending: bool) void {
        self.sort = Sort{ .field = field, .ascending = ascending };
    }

    pub fn withMethod(self: *Query, method: []const u8) void {
        self.method = method;
    }
};

pub const Manifest = struct {
    base: []const u8,
    id: ?Value,
    fields: [][]const u8,
    filter: StringHashMap(Value),
    lookups: [][]const u8,
    inlines: [][]const u8,
    page: ?usize,
    per_page: ?usize,
    sort: ?Sort,
    method: ?[]const u8,
    filter_fields: [][][]const u8,
};

pub const MutateInput = struct {
    model: []const u8,
    id: ?Value,
    data: StringHashMap([]const u8),
};

pub fn parseQueryParams(
    query_str: []const u8,
    default_per_page: usize,
    allocator: std.mem.Allocator,
) !struct { page: usize, per_page: usize, filters: StringHashMap([]const u8) } {
    var page: usize = 1;
    var per_page: usize = default_per_page;
    var filters = StringHashMap([]const u8).init(allocator);

    var parts = std.mem.splitScalar(u8, query_str, '&');
    while (parts.next()) |part| {
        if (part.len == 0) continue;
        var kv = std.mem.splitScalar(u8, part, '=');
        const k = kv.first();
        const v = percentDecode(kv.next() orelse "", allocator);

        if (std.mem.eql(u8, k, "page")) {
            page = std.fmt.parseUnsigned(usize, v, 10) catch 1;
            if (page == 0) page = 1;
        } else if (std.mem.eql(u8, k, "per_page")) {
            per_page = std.fmt.parseUnsigned(usize, v, 10) catch default_per_page;
            if (per_page == 0) per_page = 1;
        } else {
            if (v.len > 0) {
                const entry = filters.getOrPut(k) catch continue;
                if (entry.found_existing) {
                    const existing = entry.value_ptr.*;
                    const joined = try std.fmt.allocPrint(allocator, "{s},{s}", .{ existing, v });
                    entry.value_ptr.* = joined;
                } else {
                    entry.value_ptr.* = v;
                }
            }
        }
    }

    return .{ .page = page, .per_page = per_page, .filters = filters };
}

fn percentDecode(s: []const u8, allocator: std.mem.Allocator) []const u8 {
    var result = std.ArrayList(u8).init(allocator);
    var i: usize = 0;
    while (i < s.len) : (i += 1) {
        if (s[i] == '+') {
            result.append(' ') catch {};
        } else if (s[i] == '%' and i + 2 < s.len) {
            const hex = s[i + 1 .. i + 3];
            const byte = std.fmt.parseInt(u8, hex, 16) catch {
                result.append(s[i]) catch {};
                continue;
            };
            result.append(byte) catch {};
            i += 2;
        } else {
            result.append(s[i]) catch {};
        }
    }
    return result.items;
}

fn tryAllocSlice(allocator: std.mem.Allocator, items: []const []const u8) [][]const u8 {
    const result = allocator.alloc([]const u8, items.len) catch return &.{};
    @memcpy(result, items);
    return result;
}

test "parseQueryParams empty" {
    const a = std.testing.allocator;
    const r = try parseQueryParams("", 5, a);
    try std.testing.expectEqual(@as(usize, 1), r.page);
    try std.testing.expectEqual(@as(usize, 5), r.per_page);
}

test "parseQueryParams page and per_page" {
    const a = std.testing.allocator;
    const r = try parseQueryParams("page=3&per_page=10", 5, a);
    try std.testing.expectEqual(@as(usize, 3), r.page);
    try std.testing.expectEqual(@as(usize, 10), r.per_page);
}

test "parseQueryParams filters" {
    const a = std.testing.allocator;
    const r = try parseQueryParams("name=Ali&role=admin", 5, a);
    try std.testing.expectEqualStrings("Ali", r.filters.get("name").?);
    try std.testing.expectEqualStrings("admin", r.filters.get("role").?);
}

test "parseQueryParams percent encoding" {
    const a = std.testing.allocator;
    const r = try parseQueryParams("name=Ali%20Smith", 5, a);
    try std.testing.expectEqualStrings("Ali Smith", r.filters.get("name").?);
}

test "parseQueryParams repeated key joins with comma" {
    const a = std.testing.allocator;
    const r = try parseQueryParams("tags=rust&tags=web", 5, a);
    try std.testing.expectEqualStrings("rust,web", r.filters.get("tags").?);
}
